File: python/정렬/가장 큰 수.py

""" 레벨 2. 가장 큰 수 """

def solution(numbers):
    
    sorted_numbers = sorted(list(map(str, numbers)), reverse=True, key=lambda x : (x*4)[:4])
    answer = "".join(sorted_numbers)
    
    idx = 0
    while answer[idx] == '0' and idx < len(answer)-1:
        idx+=1

    return answer[idx:] 

print(solution([6, 10, 2]))             # "6210"
print(solution([3, 30, 34, 5, 9]))      # "9534330"

print(solution([0, 0, 0, 0]))      # "0"

File: Java/src/정렬/가장큰수.java
package 정렬;

import java.util.*;

public class 가장큰수 {
    
    public static void main(String[] args) {

        int[] n1 = {6, 10, 2};
        System.out.println(solution(n1));

        int[] n2 = {3, 30, 34, 5, 9};
        System.out.println(solution(n2));

        int[] n3 = {547, 54, 5};
        System.out.println(solution(n3));

        int[] n4 = {0, 0, 0, 0};
        System.out.println(solution(n4));
    }

    public static String solution(int[] numbers) {

        String[] numberString = Arrays.stream(numbers)
                                        .mapToObj(String::valueOf)
                                        .toArray(String[]::new);

        Arrays.sort(numberString, (o1, o2) -> -(o1 + o2).compareTo(o2 + o1));

        if (numberString[0].equals("0")) {
            return "0";
        }

        return String.join("", numberString);
    }
}
